chain/chain/src/resharding.rs

```diff
@@ -0,0 +1,182 @@
+/// Implementation for all resharding logic.
+/// StateSplitRequest and StateSplitResponse are exchanged across the client_actor and sync_jobs_actor.
+/// build_state_for_split_shards_preprocessing and build_state_for_split_shards_postprocessing are handled
+/// by the client_actor while the heavy resharding build_state_for_split_shards is done by sync_jobs_actor
+/// so as to not affect client.
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use near_chain_primitives::error::Error;
+use near_client_primitives::types::StateSplitApplyingStatus;
+use near_o11y::log_assert;
+use near_primitives::hash::CryptoHash;
+use near_primitives::shard_layout::{account_id_to_shard_uid, ShardLayout};
+use near_primitives::state_part::PartId;
+use near_primitives::syncing::{get_num_state_parts, STATE_PART_MEMORY_LIMIT};
+use near_primitives::types::chunk_extra::ChunkExtra;
+use near_primitives::types::{AccountId, ShardId, StateRoot};
+use near_store::split_state::get_delayed_receipts;
+use near_store::{ShardTries, ShardUId, Trie};
+
+use tracing::debug;
+
+use crate::types::RuntimeAdapter;
+use crate::Chain;
+
+#[derive(actix::Message)]
+#[rtype(result = "()")]
+pub struct StateSplitRequest {
+    pub runtime_adapter: Arc<dyn RuntimeAdapter>,
+    pub sync_hash: CryptoHash,
+    pub shard_id: ShardId,
+    pub shard_uid: ShardUId,
+    pub state_root: StateRoot,
+    pub next_epoch_shard_layout: ShardLayout,
+    pub state_split_status: Arc<StateSplitApplyingStatus>,
+}
+
+#[derive(actix::Message)]
+#[rtype(result = "()")]
+pub struct StateSplitResponse {
+    pub sync_hash: CryptoHash,
+    pub shard_id: ShardId,
+    pub new_state_roots: Result<HashMap<ShardUId, StateRoot>, Error>,
+}
+
+fn apply_delayed_receipts<'a>(
+    tries: &ShardTries,
+    orig_shard_uid: ShardUId,
+    orig_state_root: StateRoot,
+    state_roots: HashMap<ShardUId, StateRoot>,
+    account_id_to_shard_id: &(dyn Fn(&AccountId) -> ShardUId + 'a),
+) -> Result<HashMap<ShardUId, StateRoot>, Error> {
+    let orig_trie_update = tries.new_trie_update_view(orig_shard_uid, orig_state_root);
+
+    let mut start_index = None;
+    let mut new_state_roots = state_roots;
+    while let Some((next_index, receipts)) =
+        get_delayed_receipts(&orig_trie_update, start_index, STATE_PART_MEMORY_LIMIT)?
```

Comment:
cool syntax though hard to parse at first sight

Reply:
Haha! I'd done the same in `core/store/src/trie/from_flat.rs`